### ISSUE TITLE
'Chicago' citation style does not work properly

### ISSUE THREAD

**Issue by Joeran** [Docear/Docear4Word#17]
There are a few issues with the popular Chicago citation style.
1. We included the wrong one in the default list of citation styles that is installed with Docear4Word. Hence, remove the current style "Chicago Manual of Style (note, annotated bibliography)" from the default list of the styles that Docear4Word installs. Instead, add the style "Chicago Manual of Style 16th edition (author-date)" to the default list https://www.zotero.org/styles/chicago-author-date
2. The Chicago style, but also some other styles (e.g. "American Institute of Physics"), spaces and words like "of" and "and" are removed from titles and journal names in the bibliography. This is not how it should be. 
   ![image](https://cloud.githubusercontent.com/assets/2410381/3213977/95f397e8-efa3-11e3-960d-3f229c71f781.png)

**Comment by hudcap** [comment]
I believe the spaces are being removed due to an incorrect processing of the CSL attribute text-case="title"

I ran into this bug while customizing my own style and removing this attribute rendered the spaces correctly.

**Comment by DaveLaur** [comment]
I also came across this problem for a CSL file for modern-humanities-research-association.csl, so had a look at the source code.  The bug fix is very simple...

On lines 12362 and 12364 of citeproc.js, a space is missing in the join argument string.  They should be:

        lst[i] = words.join(" ");

    doppel.string = lst.join(" ");

respectively.  This file is found in the JavaScript directory under the Docear4word programme directory.  Editing the Javascript file sorts the problem out without any recompilation of code required, so can be done by an end-user (though, obviously it would be better to get the source corrected - I'm new to Github so don't know how to do this)!

Hopefully someone who knows what they are doing can take this and (a) check that the correction is right, and (b) update the source files.

**Comment by Joeran** [comment]
hello,

we are not the authors of the citeproc.js. the citeproc project can be found here https://bitbucket.org/fbennett/citeproc-js/wiki/Home hence any changes should be suggested there (and i guess they might have done the changes already).

could you download the latest citeproc.js https://bitbucket.org/fbennett/citeproc-js/src/f73704ce6679dc03117b5c893a4fdbb1ea066e98/citeproc.js?at=default and replace the old one in the Docear directory with the new one? and let me know if this fixes the problems.

**Comment by DaveLaur** [comment]
I’m afraid not – it just breaks the whole thing – the docear4word add-in no longer works.  (Is citeproc.js part of a package, and the other files would also be required)?

 

Also, quickly looking through the script, it has the same join lines in the “title” case formatting, so I suspect will do exactly the same thing on the formatting…

 

Dave.

 

***************************************************

Dave Laurenson, Deputy Director of Teaching

The School of Engineering, The University of Edinburgh,

The King's Buildings, Edinburgh EH9 3JL, Scotland, UK

 

Phone: +44 (0) 131 650 5579    Fax: +44 (0) 131 650 6554

Email:  <mailto:Dave.Laurenson@ed.ac.uk> Dave.Laurenson@ed.ac.uk             <http://www.eng.ed.ac.uk/> www.eng.ed.ac.uk

**Comment by Joeran** [comment]
> Is citeproc.js part of a package, and the other files would also be required

Not really. I don't know why the current citeproc.js causes such problems. Anyway, I created a FAQ entry with your proposed solution http://www.docear.org/faqs/why-are-spaces-removed-in-the-bibliography/

**Comment by DaveLaur** [comment]
Unfortunately when I looked at it more carefully, it is missing out the small words that should not be capitalized. I suspect the code that does the splitting of words isn't doing what the author of citeproc expected.  I'm going to dig into it further when I get a chance.

Dave.

On 27 June 2015 12:24:03 BST, Joeran <notifications@github.com> wrote:
>> Is citeproc.js part of a package, and the other files would also be
>required
>
>Not really. I don't know why the current citeproc.js causes such
>problems. Anyway, I created a FAQ entry with your proposed solution
>http://www.docear.org/faqs/why-are-spaces-removed-in-the-bibliography/
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/Docear/Docear4Word/issues/17#issuecomment-116013829

-- 
Sent from my Xperia smartphone with K-9 Mail. Please excuse my brevity.
The University of Edinburgh is a charitable body, registered in
Scotland, with registration number SC005336.

**Comment by DaveLaur** [comment]
I have looked into this properly – the code in that particular section has to be about the most badly written code that I’ve seen for some time!  It is trying to be too clever, tying itself in knots, and making some very stupid basic errors, like only processing every other word/clause.  I’ve rewritten code that appears to work quite happily, but is significantly simplified from what was there.

 

I’d suggest that users comment out lines 12336 to 12364 inclusive by inserting /* at the start of line 12336, and */ at the end of 12364.  Then either before, or after this, insert the following:

 

    // Code replaced by D. Laurenson to process with docear4word

    //

    // In this code, no attempt is made to ensure that sentence beginnings

    // are made uppercase.  Without searching for all possible abbreviations

    // such a check would be only partial at best.  The faulty code

    // attempted to captialise all words following :, ? or !, but not .

    //

    // Start by splitting the string according to spaces only

    var words = str.split(/\s/);

    // For each word, check against the non-capitalise list, and if not in the list, then captialise it

    for (i=0,ilen=words.length;i<ilen;i+=1) {

                if (!words[i].match(state.locale[state.opt.lang].opts["skip-words-regexp"])) {

                    words[i] = capitalise(words[i]);

                }

    }

    // Reconstruct the string with the words in title case

    doppel.string = words.join(" ");

 

 

 

As you can see, it is relatively straightforward code – nothing fancy, but that appears to be in its favour!  I’ve also taken the liberty of putting in comments – something the original author didn’t bother with.  If you think these are superfluous, please just remove any lines preceded by //.

 

This time I’ve checked it with a number of very specific examples to stretch its capabilities, using escaped characters, etc., to make sure that it works.

 

All te best,

 

Dave.

 

***************************************************

Dave Laurenson, Deputy Director of Teaching

The School of Engineering, The University of Edinburgh,

The King's Buildings, Edinburgh EH9 3JL, Scotland, UK

 

Phone: +44 (0) 131 650 5579    Fax: +44 (0) 131 650 6554

Email:  <mailto:Dave.Laurenson@ed.ac.uk> Dave.Laurenson@ed.ac.uk             <http://www.eng.ed.ac.uk/> www.eng.ed.ac.uk

**Comment by faph** [comment]
It might be more effective if an issue is raised with siteproc.js here: https://bitbucket.org/fbennett/citeproc-js/issues

Docear4Word can then simply make a new release with an updated citeproc library.

**Comment by DaveLaur** [comment]
I have raised this with the author of citeproc (Frank Bennett).  It turns out that this issue is the result of Javascript incompatibilities in the function used to split strings, and he is looking to put a fix in for this.

 

Docear4word doesn’t ship with the latest citeproc.js, and when I tried it, it didn’t seem to like it either, but I think that is a docear4word problem.

 

Dave.

 

***************************************************

Dave Laurenson, Deputy Director of Teaching

The School of Engineering, The University of Edinburgh,

The King's Buildings, Edinburgh EH9 3JL, Scotland, UK

 

Phone: +44 (0) 131 650 5579    Fax: +44 (0) 131 650 6554

Email:  <mailto:Dave.Laurenson@ed.ac.uk> Dave.Laurenson@ed.ac.uk             <http://www.eng.ed.ac.uk/> www.eng.ed.ac.uk

**Comment by DmitriyC** [comment]
As a potential workaround for this issue, the current beta version of JabRef (2.11b4) has support for converting entries to title-case capitalization (although the function can only be applied to one entry at a time, unfortunately).  To avoid the citeproc.js bug, users can then simply remove title text-case formatting from the citation style.

Dave, thank you for the suggestions.  There is a small copy-paste error from the original code (i vs. k), but it works well.  I've extended your code to cover a few punctuation marks (hyphenation, slashes, etc.), but obviously there are still a lot of limitations.  I can't attach files, but the code is located at http://pastie.org/10541668.  My programming skill is still fairly basic and I'm entirely new to JavaScript, but the code seems to work fine.  The goal was to minimize the amount of final editing the end user needs to make to their bibliography.
